input: add a gpio based keyboard matrix driver

fabiobaltieri opened the pull request:
Initial gpio based driver for the now generalized keyboard scanning library. Should be useful to tweak and troubleshoot it the generic keyboard scanning code on any random devboard. Just tested it on a random board shorting pins out, got some dedicated hardware to do more testing on the way. Prob deserves some better documentation but I have more features and options for this in the pipeline so I think I'd work on a nice doc page at a later stage.

---

Add a GPIO based keyboard matrix driver using the generic keyboard matrix code.

fabiobaltieri committed:
input: add a gpio based keyboard matrix driver

Add a GPIO based keyboard matrix driver using the generic keyboard
matrix code.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

gmarull commented on tests/drivers/build_all/input/app.overlay:
```diff
@@ -26,6 +26,17 @@
+			col-gpios = <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>;
+			row-size = <2>;
+			col-size = <3>;
```
couldn't these be inferred from len(row-gpios) and len(col-gpios)?

fabiobaltieri replied:
not really, they are defined in the generic code while the gpios one are in this specific implementation

gmarull replied:
I mean, the driver could make these optional in this particular case?

fabiobaltieri replied:
...unless I duplicate the initializer and make a version that takes row and col as an argument, how does that sound?

gmarull replied:
up to you, just thought these are a bit redundant in this case.

fabiobaltieri replied:
alright how's it look now? more macros more better

fabiobaltieri committed:
input: kbd: make row and col size optional add extra macro

Change the row-count and col-count to be optional in the generic
binding, add a second pair of macro to allow the implementation to
specify the numbers from a different property.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on drivers/input/input_gpio_kbd_matrix.c:
```diff
@@ -0,0 +1,199 @@
+												\
+	static const struct gpio_kbd_matrix_config gpio_kbd_matrix_cfg_##n = {			\
+		.common = INPUT_KBD_MATRIX_DT_INST_COMMON_CONFIG_INIT_ROW_COL(			\
+			0, gpio_kbd_matrix_api,							\
```
sneaky `0`

gmarull commented on dts/bindings/input/gpio-kbd-matrix.yaml:
```diff
@@ -0,0 +1,40 @@
+          col-gpios = <&gpio0 2 GPIO_ACTIVE_LOW>,
+                      <&gpio0 3 GPIO_ACTIVE_LOW>,
+                      <&gpio0 4 GPIO_ACTIVE_LOW>;
+          row-size = <2>;
+          col-size = <3>;
+          no-ghostkey-check;
+  };
+
+compatible: "gpio-kbd-matrix"
+
+include: kbd-matrix-common.yaml
```
```suggestion
          no-ghostkey-check;
  };

compatible: "gpio-kbd-matrix"

include: kbd-matrix-common.yaml
```

maybe filter-out row-size/col-size when including?

fabiobaltieri replied:
done, good idea, also done

fabiobaltieri commented on drivers/input/input_gpio_kbd_matrix.c:
```diff
@@ -0,0 +1,199 @@
+
+struct gpio_kbd_matrix_data {
+	struct input_kbd_matrix_common_data common;
+	int last_col_state;
```
this should really be a uint32

fabiobaltieri commented:
@gmarull sneaked in one more tweak post approval, the col limit of 32 is beacuse of the last state in data, made in an `uint32_t` explicitly.

fabiobaltieri committed:
input: kbd_matrix: make the api field a pointer

The API field of input_kbd_matrix_common_config should have been a
pointer from the start, clang-16 caught this with a compiler warning.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Turns out I had messed up the common kbd config struct and had the api struct itself rather than a point to it, clang-16 fired up a (fairly cryptic) warning about it, guess it was working by copying the structure fields at init time (didn't even know it was possible). Slipped in another commit to fix that as well.

fabiobaltieri committed:
doc: input: add a input_kbd_matrix group, fix few missing doc entries

Add a input_kbd_matrix doxygen group and add this to the other Input
APIs page, add few missing argument documentation entries.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on tests/drivers/build_all/input/app.overlay:
```diff
@@ -26,6 +26,15 @@
+				    <&gpio0 0 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
+			col-gpios = <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>,
+				    <&gpio0 0 GPIO_ACTIVE_LOW>;
```
Ok, I  changed it to sequential numbers. Spotted some other incoherences in this file btw, I'll fix the rest separately.

fabiobaltieri committed:
input: kbd_matrix: move the header to the project wide include directory

Move the input_kbd_matrix.h header out of drivers/ and into include/,
this allows external drivers to use it and doxygen to pick it up.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Ah, doxygen is missing the file, guess it should have been in `include/` anyway, moved, let's see if it like it better.

keith-zephyr commented on dts/bindings/input/gpio-kbd-matrix.yaml:
```diff
@@ -0,0 +1,42 @@
+    type: phandle-array
+    required: true
+    description: |
+      GPIO for the keyboard matrix rows, must match the count in row-size,
+      supports up to 8 rows, all row GPIO pins must have interrupt support.
```
Drop the "match row-size" language.
```suggestion
      GPIO for the keyboard matrix rows. All row GPIO pins must have interrupt support.
```

fabiobaltieri replied:
Done, I kept the note about the limit of 8 as that still applies, it's enforced in the common code.

keith-zephyr commented on dts/bindings/input/gpio-kbd-matrix.yaml:
```diff
@@ -0,0 +1,42 @@
+    type: phandle-array
+    required: true
+    description: |
+      GPIO for the keyboard matrix columns, must match the count in col-size,
+      supports up to 32 columns.
```
Drop "match col-size" languate.

fabiobaltieri replied:
Thanks! Done, added a note on column gpio driving as well.

fabiobaltieri commented:
Ok I bumped into few more doxygen issues after checking https://builds.zephyrproject.io/zephyr/pr/65117/docs/services/input/index.html#keyboard-matrix-api-reference, it should be ok now.

Could probably do a better job at rewriting the commit history but I'd rather spare myself the rebase conflicts, this PR took few unexpected turns, can't complain about the end result though, many things needed fixing.

Thanks for your patience on the reviews.
